mlc_llm/relax_model/llama.py

```diff
@@ -460,14 +456,21 @@ def forward(
             attention_mask=attention_mask,
             all_seq_len_shape=all_seq_len_shape,
         )
+        if self.self_attn.num_shards > 1:
+            residual = nn.emit(residual / R.const(self.self_attn.num_shards, dtype="float16"))
```

Comment:
sorry i dont fully understand this change. could you briefly explain?

Reply:
previously the pattern is matmul->allreduce->residual_add->xxx, which leaves the add kernel not fused. Now we reorder it into matmul->residual_add->allreduce->xxx, so everything can be fused.

Reply:
and we need to divide the residual tensor by shard num to ensure the arithmetic correctness of the reorder